src/ILCompiler.Compiler/src/IL/Stubs/StartupCode/StartupCodeMainMethod.cs

```diff
@@ -89,9 +89,20 @@ public override MethodIL EmitIL()
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(startup.GetKnownMethod("GetMainMethodArguments", null)));
             }
             codeStream.Emit(ILOpcode.call, emitter.NewToken(_mainMethod));
+
+            // If the main method returns void, return the latched exit code set by Environment.ExitCode.
             if (_mainMethod.Signature.ReturnType.IsVoid)
             {
-                codeStream.EmitLdc(0);
+                MetadataType environmentAugments = Context.SystemModule.GetType("Internal.Runtime.Augments", "EnvironmentAugments");
```

Comment:
What is in ProjectN will need revamping for ns20. I am not able to tell what the sharing will actually look like. I would optimize for aesthetics and ns20 in CoreRT with this change.

The shutdown sequence that we are going to need for ns20 will actually need to look like the following - the closer we are to it, the better:

```
if (!_mainMethod.Signature.ReturnType.IsVoid)
{
    emit StartupCodeHelpers.SetLatchedExitCode(<main return value>);
}

emit return StartupCodeHelpers.Shutdown();
```

For example, this program should have exit code 5:

```
using System;

class Program
{
    static int Main(string[] args)
    {
        AppDomain.CurrentDomain.ProcessExit += CurrentDomain_ProcessExit;
        return 12;
    }

    private static void CurrentDomain_ProcessExit(object sender, EventArgs e)
    {
        Environment.ExitCode = 5;
    }
}
```